identity-account/src/updates/update.rs

```diff
@@ -0,0 +1,465 @@
+
+  let mut state = IdentityState::new(document);
+
+  // Store the generations at which the method was added
+  state.store_method_generations(method_fragment);
+
+  Ok((did_lease, state))
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum Update {
+  CreateMethod {
+    scope: MethodScope,
+    type_: MethodType,
+    fragment: String,
+    method_secret: Option<MethodSecret>,
+  },
+  DeleteMethod {
+    fragment: String,
+  },
+  AttachMethod {
+    fragment: String,
+    relationships: Vec<MethodRelationship>,
+  },
+  DetachMethod {
+    fragment: String,
+    relationships: Vec<MethodRelationship>,
+  },
+  CreateService {
+    fragment: String,
+    type_: String,
+    endpoint: ServiceEndpoint,
+    properties: Option<Object>,
+  },
+  DeleteService {
+    fragment: String,
+  },
+}
+
+impl Update {
+  pub(crate) async fn process(self, did: &IotaDID, state: &mut IdentityState, storage: &dyn Storage) -> Result<()> {
+    debug!("[Update::process] Update = {:?}", self);
+    trace!("[Update::process] State = {:?}", state);
+    trace!("[Update::process] Store = {:?}", storage);
+
+    match self {
+      Self::CreateMethod {
+        type_,
+        scope,
+        fragment,
+        method_secret,
+      } => {
+        let location: KeyLocation = state.key_location(type_, fragment)?;
+
+        // The key location must be available.
+        // TODO: config: strict
+        ensure!(
+          !storage.key_exists(did, &location).await?,
+          UpdateError::DuplicateKeyLocation(location)
+        );
+
+        // The verification method must not exist.
+        ensure!(
+          state
+            .document()
+            .resolve_method(location.fragment().identifier())
+            .is_none(),
+          UpdateError::DuplicateKeyFragment(location.fragment().clone()),
+        );
+
+        let public: PublicKey = if let Some(method_private_key) = method_secret {
+          insert_method_secret(storage, did, &location, type_, method_private_key).await
+        } else {
+          storage.key_new(did, &location).await
+        }?;
+
+        let method: IotaVerificationMethod =
+          IotaVerificationMethod::from_did(did.to_owned(), KeyType::Ed25519, &public, location.fragment().name())?;
+
+        state.store_method_generations(location.fragment().clone());
+
+        // We can ignore the result: we just checked that the method does not exist.
+        state.document_mut().insert_method(method, scope);
+      }
+      Self::DeleteMethod { fragment } => {
+        let fragment: Fragment = Fragment::new(fragment);
+
+        // The verification method must exist
+        ensure!(
+          state.document().resolve_method(fragment.identifier()).is_some(),
+          UpdateError::MethodNotFound
+        );
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
+        let core_method_url: CoreDIDUrl = CoreDIDUrl::from(method_url.clone());
```

Comment:
I think so, yes. The fragment is user-input and could contain invalid or reserved characters. `Fragment` is separate from `DIDUrl` (annoyingly) and doesn't validate the contents.